### PR TITLE
nixos/alps: init module

### PR DESCRIPTION
**Pull request by madonius**
###### Description of changes

Added service definition for alps

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by madonius** [comment]
What is missing:
[ ] Service restart does not work
[ ] Systemd target dependency is not correct

File: pkgs/servers/alps/default.nix
Comment (NickCao) [inline pkgs/servers/alps/default.nix:28]:
```suggestion
  ldflags = [ "-X main.themesPath=${src}/themes" ];
```

Comment (madonius) [reply pkgs/servers/alps/default.nix]: Thanks for the feedback :)

File: pkgs/servers/alps/default.nix
Comment (madonius) [reply pkgs/servers/alps/default.nix:25]: The plugins seem to be working fine, tbh. As I am not that familiar with alps, I am not sure whether I would know if something wasn't working.

Comment (NickCao) [reply pkgs/servers/alps/default.nix]: Or we just leave it as as (as we have been distributing a broken alps anyway).

**Comment by NickCao** [comment]
To acoid cluttering this pr with the packaging fixes, I extracted that part into a separate pr: https://github.com/NixOS/nixpkgs/pull/180699

File: nixos/modules/services/web-apps/alps.nix
Comment (NickCao) [inline nixos/modules/services/web-apps/alps.nix:35]: This whole part together with smtps below can be simplified into a single option `upstreams`, a list of string, to align with the upstream flags.

File: nixos/modules/services/web-apps/alps.nix
Comment (NickCao) [inline nixos/modules/services/web-apps/alps.nix:11]: This too can be simplified as a single option `listenAddress`, which allows for neat shorthand like `:8080`.

File: nixos/modules/services/web-apps/alps.nix
Comment (NickCao) [inline nixos/modules/services/web-apps/alps.nix:82]:
```suggestion
```

File: nixos/modules/services/web-apps/alps.nix
Comment (NickCao) [inline nixos/modules/services/web-apps/alps.nix:94]:
```suggestion
```

File: nixos/modules/services/web-apps/alps.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/alps.nix:40]:
```suggestion
          The IMAP server port
```

File: nixos/modules/services/web-apps/alps.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/alps.nix:31]:
```suggestion
        The frontend's theme to use.
```

File: nixos/modules/services/web-apps/alps.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/alps.nix:23]:
```suggestion
        The IP the service should listen to.
```

File: nixos/modules/services/web-apps/alps.nix
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/alps.nix:15]:
```suggestion
        TCP port the service should listen to.
```

**Comment by booklearner** [comment]
Thanks for adding the reviews,  and thanks adding the service and adding yourself as a maintainer!

This LGTM after feedback from other reviews is applied.

File: nixos/modules/services/web-apps/alps.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/alps.nix:35]: > a list of string, to align with the upstream flags.

be careful with that, we need either a single argument or two, not more, so we'd need some validation here.

File: nixos/modules/services/web-apps/alps.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/alps.nix:85]: What exactly? If you're referring to the package, there are overlays for that.

File: nixos/modules/services/web-apps/alps.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/alps.nix:49]: I think that `imaps` was actually reasonable to state that this only works with that.

File: pkgs/servers/alps/default.nix
Comment (Ma27) [reply pkgs/servers/alps/default.nix]: There's no point in merging a non-functional module though.

File: nixos/modules/services/web-apps/alps.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/alps.nix:11]: * The "neat shorthand" will AFAIU only appear in the end-result, so it's not that interesting, isn't it?
* I think it's a reasonable use-case to change the listenIP while wanting to keep the default port, hence two options.

File: nixos/modules/services/web-apps/alps.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/alps.nix:94]: This should be set back to /var/lib/alps to ensure that anything written into cwd by alps is persisted in the state-directory.

File: pkgs/servers/alps/default.nix
Comment (Ma27) [reply pkgs/servers/alps/default.nix:24]: OT, but why do we set this on every single package with ld flags? Seems like something `buildGoModule` should take care of rather than every author having to remember this.

File: nixos/modules/services/web-apps/alps.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/alps.nix:80]: Isn't network-online.target reached after network.target?

File: pkgs/servers/alps/default.nix
Comment (NickCao) [reply pkgs/servers/alps/default.nix:24]: Adding `-s -w` by default would break some programs, see https://github.com/NixOS/nixpkgs/issues/177698

File: nixos/modules/services/web-apps/alps.nix
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/alps.nix:80]: Yes but usually both are added

**Comment by Ma27** [comment]
@madonius can we help you with pushing this forward? :)

File: nixos/modules/services/web-apps/alps.nix
Comment (revol-xut) [reply nixos/modules/services/web-apps/alps.nix:49]: The config he produces only supports IMAP with SSL so the imaps makes sense because its the secure version.

My suggestion would be: `The address of the secure IMAP server`.

File: nixos/modules/services/web-apps/alps.nix
Comment (madonius) [reply nixos/modules/services/web-apps/alps.nix:94]: The matter with that is that the software is searching for the themes and the plugins in its working directory. (Yes I know, this should™ be configurable)
Copying the respective directories to `/var/lib/` did not work out because after restarting the service, systemd's ephemeral user changed and the service did ofcourse not start due to missing permissions on the directories.
A solution to that would be appreciated

File: nixos/modules/services/web-apps/alps.nix
Comment (madonius) [reply nixos/modules/services/web-apps/alps.nix:49]: I opted to use the formulation with IMAPs, I hope this is agreeable.

File: nixos/modules/services/web-apps/alps.nix
Comment (madonius) [reply nixos/modules/services/web-apps/alps.nix:82]: Sorry for lacking the understanding. Why is coreutils not needed here?

File: nixos/modules/services/web-apps/alps.nix
Comment (revol-xut) [reply nixos/modules/services/web-apps/alps.nix:11]: I also prefer to split it into options so its easier to use from the user site. Less opportunity to miss configure.

File: nixos/modules/services/web-apps/alps.nix
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/alps.nix:81]: coreutils is in the default PATH

**Review by SuperSandro2000 — APPROVED** [review]

**Review by booklearner — APPROVED** [review]

File: nixos/modules/services/web-apps/alps.nix
Comment (hmenke) [inline nixos/modules/services/web-apps/alps.nix:57]: Typo? The default SMTPS port is 465.

Comment (booklearner) [reply nixos/modules/services/web-apps/alps.nix:57]: Good catch, thanks!

I opened a PR to fix this: https://github.com/NixOS/nixpkgs/pull/199905

**Pull request merged**
